registry/storage/driver/filesystem/driver.go

```diff
@@ -291,6 +292,50 @@ func (d *driver) Move(ctx context.Context, sourcePath string, destPath string) e
 	}
 
 	err := os.Rename(source, dest)
+	if err != nil {
+		// Copy then delete fall back
+		linkErr := err.(*os.LinkError)
+		if errno, ok := linkErr.Err.(syscall.Errno); ok && errno == syscall.EXDEV || os.IsExist(err) {
+			err = d.copyFile(source, dest)
+			// If we just moved an uploaded blob, the PurgeUploads loop probably removed it already,
+			// therefore we must check if the source file still exists before trying to delete it
+			if err = os.RemoveAll(source); err != nil {
+				return err
+			}
+		}
+	}
+	return err
+}
+
+// Copy a file to another. If dst file exists, it will be overwritten
+func (d *driver) copyFile(src, dst string) (err error) {
+	in, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer in.Close()
+	// Allocate tmp file with unique name, otherwise multiple processes may write to the same file
+	out, err := ioutil.TempFile(path.Dir(dst), path.Base(dst))
+	if err != nil {
+		return err
+	}
+	if _, err = io.Copy(out, in); err != nil {
```

Comment:
The logic here still doesn't quite right yet.

We really want something like https://github.com/docker/docker/blob/master/pkg/ioutils/fswriters.go, but we don't want to import docker/docker here. There should likely be a defer that happens before the copy which will handle the cleanup if `err` is not nil and always closes. The call to `RemoveAll` should be skipped if rename passed, in fact if rename passed the close should still happen but probably shouldn't return an error (after rename passed this function succeeded and returning an error might be unexpected to the caller).